osquery/filesystem/linux/proc.h

```diff
@@ -0,0 +1,248 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#pragma once
+
+#include <unordered_map>
+
+#include <arpa/inet.h>
+#include <linux/limits.h>
+#include <unistd.h>
+
+#include <boost/filesystem.hpp>
+
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+
+namespace osquery {
+extern const char* kLinuxProcPath;
+
+struct ProcessSocket final {
+  std::string socket;
+  int family;
+  int protocol;
+
+  std::string local_address;
+  std::uint16_t local_port;
+
+  std::string remote_address;
+  std::uint16_t remote_port;
+
+  std::string unix_socket_path;
+
+  int fd;
+};
+
+// Linux proc protocol define to net stats file name.
+const std::map<int, std::string> kLinuxProtocolNames = {
+    {IPPROTO_ICMP, "icmp"},
+    {IPPROTO_TCP, "tcp"},
+    {IPPROTO_UDP, "udp"},
+    {IPPROTO_UDPLITE, "udplite"},
+    {IPPROTO_RAW, "raw"},
+};
+
+using ProcessNamespaceList = std::map<std::string, ino_t>;
+
+Status procGetProcessNamespaces(const std::string& process_id,
+                                ProcessNamespaceList& namespace_list);
+
+Status procReadDescriptor(const std::string& process,
+                          const std::string& descriptor,
+                          std::string& result);
+
+std::string procDecodeAddressFromHex(const std::string& encoded_address,
+                                     int family);
+
+unsigned short procDecodePortFromHex(const std::string& encoded_port);
+
+Status procSocketInodeToFdMap(
+    const std::string& process,
+    std::unordered_map<std::string, std::string>& inode_to_fd_map);
+
+template <typename UserData>
+Status procProcesses(bool (*callback)(const std::string&, UserData),
```

Comment:
Have you considered using `std::bind` to generate a callback function without needing to template this?

Reply:
Hey @fmanco! Depending on the compiler, using std::bind is much slower; since the callback is supposed to be called many times, I ended up avoiding it.